src/query/storage/m3/cluster_resolver.go

```diff
@@ -0,0 +1,299 @@
+	}
+
+	retention := unaggregated.Options().Attributes().Retention
+	unaggregatedStart := now.Add(-1 * retention)
+	if unaggregatedStart.Before(start) || unaggregatedStart.Equal(start) {
+		return unaggregatedNamespaceDetails{
+			clusterNamespace: unaggregated,
+			satisfies:        fullySatisfiesRange,
+		}
+	}
+
+	return unaggregatedNamespaceDetails{
+		clusterNamespace: unaggregated,
+		retention:        retention,
+		satisfies:        partiallySatisfiesRange,
+	}
+}
+
+// resolveClusterNamespacesForQuery returns the namespaces that need to be
+// fanned out to depending on the query time and the namespaces configured.
+func resolveClusterNamespacesForQuery(
+	now time.Time,
+	clusters Clusters,
+	start time.Time,
+	end time.Time,
+	opts *storage.FanoutOptions,
+) (queryFanoutType, ClusterNamespaces, error) {
+	// First check if the unaggregated cluster can fully satisfy the query range.
+	// If so, return it and shortcircuit, as unaggregated will necessarily have
+	// every metric.
+	unaggregated := resolveUnaggregatedNamespaceForQuery(now, start,
+		clusters.UnaggregatedClusterNamespace(), opts)
+	if unaggregated.satisfies == fullySatisfiesRange {
+		return namespaceCoversAllQueryRange,
+			ClusterNamespaces{unaggregated.clusterNamespace},
+			nil
+	}
+
+	// The filter function will drop namespaces which do not cover the entire
+	// query range from contention.
+	//
+	// NB: if fanout aggregation is forced on, the filter instead forces clusters
+	// that do not cover the range to be set as partially aggregated.
+	filterFunc := func(namespace ClusterNamespace) bool {
+		// Include only if can fulfill the entire time range of the query
+		clusterStart := now.Add(-1 * namespace.Options().Attributes().Retention)
+		return !clusterStart.After(start)
+	}
+
+	// Filter aggregated namespaces by filter function and options.
+	var r reusedAggregatedNamespaceSlices
+	r = aggregatedNamespaces(clusters.ClusterNamespaces(), r, filterFunc, opts)
+
+	// If any of the aggregated clusters have a complete set of metrics, use
+	// those that have the smallest resolutions, supplemented by lower resolution
+	// partially aggregated metrics.
+	if len(r.completeAggregated) > 0 {
+		sort.Stable(ClusterNamespacesByResolutionAsc(r.completeAggregated))
+		// Take most granular complete aggregated namespace.
+		result := r.completeAggregated[:1]
+		completedAttrs := result[0].Options().Attributes()
+		// Also include any finer grain partially aggregated namespaces that
+		// may contain a matching metric.
+		for _, n := range r.partialAggregated {
+			if n.Options().Attributes().Resolution < completedAttrs.Resolution {
+				// More granular resolution.
+				result = append(result, n)
+			}
+		}
+
+		return namespaceCoversAllQueryRange, result, nil
+	}
+
+	// No need to regenerate aggregated namespace without a filter if aggregated
+	// fanout is force enabled.
+	if opts.FanoutAggregated != storage.FanoutForceEnable {
```

Comment:
Hm isn't the default to enable fanning out to aggregated namespaces in general?  

i.e. if `ForceEnable` and `Default` should both always return aggregated namespaces, and for `FanoutAggregatedOptimized` that should just determine whether to include or not include namespaces with `downsampleOpts.All` into the complete or partial bucket.

So I would've thought for `opts.FanoutAggregated == FanoutForceDisable` we would just always return the unaggregated namespace, and no namespaces if there is no unaggregated namespace at all.  And then we shouldn't check this flag any later I would've thought?